### ISSUE TITLE
IPv6 support in EKS bootstrap script is broken when setting B64_CLUSTER_CA and APISERVER_ENDPOINT

### ISSUE THREAD

**Issue by archoversight** [terraform-aws-modules/terraform-aws-eks#1958]
## Description

Please see:

https://github.com/awslabs/amazon-eks-ami/issues/887

## Versions
```
% terraform --version
Terraform v1.1.7
on darwin_amd64
+ provider registry.terraform.io/hashicorp/aws v4.6.0
+ provider registry.terraform.io/hashicorp/cloudinit v2.2.0
+ provider registry.terraform.io/hashicorp/helm v2.4.1
+ provider registry.terraform.io/hashicorp/null v3.1.1
+ provider registry.terraform.io/hashicorp/random v3.1.2
+ provider registry.terraform.io/hashicorp/time v0.7.2
+ provider registry.terraform.io/hashicorp/tls v3.1.0
```

Module version: ``` "~> 18.8.1"```

## Reproduction

Deploy a cluster with IPv6 using `self_managed_node_groups` with all defaults. Cluster will attempt to set DNS IP address to `127.x.x.x` address instead of the IPv6 address.

Deploying a custom bootstrap template:

```
      user_data_template_path = "${path.module}/eks-bootstrap.tpl"
```

That removes the `B64_CLUSTER_CA` and `APISERVER_ENDPOINT` can help work around the problem, as long as you also deploy the aws cli v2:

```
%{ if enable_bootstrap_user_data ~}
#!/bin/bash
set -e
%{ endif ~}
${pre_bootstrap_user_data ~}
%{ if length(cluster_service_ipv4_cidr) > 0 ~}
export SERVICE_IPV4_CIDR=${cluster_service_ipv4_cidr}
%{ endif ~}
%{ if enable_bootstrap_user_data ~}
    echo "Fetching AWS CLIv2"
    (
      cd /tmp
      curl "https://awscli.amazonaws.com/awscli-exe-linux-x86_64.zip" -o "awscliv2.zip"
      unzip awscliv2.zip
      ./aws/install
      rm -rf /bin/aws
      ln -s /usr/local/bin/aws /bin/aws
    )
/etc/eks/bootstrap.sh ${cluster_name} ${bootstrap_extra_args}
${post_bootstrap_user_data ~}
%{ endif ~}
```

## Expected behavior

Disable the setting of the `B64_CLUSTER_CA`/`APISERVER_ENDPOINT` when using IPv6 so that this just works out of the box, or somehow get the service address range from EKS and put that information into the appropriate flags for `/etc/eks/bootstrap.sh` to function correctly. (although I can't find any output from the terraform EKS resource that it would output that IPv6 service address).

## Actual behavior

Cluster comes up in mixed IPv4/IPv6 state and is very broken.

## Additional context

This should be fixed upstream, but AWS ain't exactly speedy in fixing their stuff... see https://github.com/awslabs/amazon-eks-ami/issues/887 and https://github.com/awslabs/amazon-eks-ami/issues/839

**Comment by marcuz** [comment]
This was a blocker for me and I've worked it around adding:

```
bootstrap_extra_args = "--ip-family ipv6 --service-ipv6-cidr $(aws eks describe-cluster --region=${var.region} --name=${var.cluster_name} --output=text --query 'cluster.{serviceIpv6Cidr: kubernetesNetworkConfig.serviceIpv6Cidr}')"
```

in my `self_managed_node_groups` definition.

Not elegant, but does the trick and pods are now starting with the correct nameserver in `/etc/resolv.conf`:

```
# cat /etc/resolv.conf 
nameserver fdcc:9f49:63e5::a
search default.svc.cluster.local svc.cluster.local cluster.local eu-west-2.compute.internal
options ndots:5
```

**Comment by bryantbiggs** [comment]
this should be resolved on the next EKS AMI release per https://github.com/awslabs/amazon-eks-ami/pull/860

**Comment by archoversight** [comment]
I am not sure that I agree that it will be fixed with that issue. After that issue the bootstrap script still assumes IPv4 if the `B64_CLUSTER_CA` and `APISERVER_ENDPOINT` endpoint are set, because this line:

https://github.com/literalice/amazon-eks-ami/blob/a2d4c4c924e9576e57cf8aaaa6e57a25015b8e5f/files/bootstrap.sh#L331

Falls through, and then:

https://github.com/literalice/amazon-eks-ami/blob/a2d4c4c924e9576e57cf8aaaa6e57a25015b8e5f/files/bootstrap.sh#L370

This sets the family to IPv4.

**Comment by bryantbiggs** [comment]
you just need to set `IP_FAMILY` = `ipv6` now and it should work now that issue 860 is resolved

**Comment by bryantbiggs** [comment]
closing now that the upstream issue is resolved. 

tl;dr - users can set `--ip-family` and `--service-ipv6-cidr` (or `IP_FAMILY` and `SERVICE_IPV6_CIDR` env vars) to enable IPv6 support. No changes to the module required

**Comment by archoversight** [comment]
@bryantbiggs how do I fetch the `SERVICE_IPV6_CIDR` from terraform to pass it into the Cloud-Init script?

**Comment by archoversight** [comment]
> you just need to set `IP_FAMILY` = `ipv6` now and it should work now that issue 860 is resolved

This is incorrect. You also need to set either the `--service-ipv6-cidr` or the `--dns-cluster-ip`.

https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/eks_cluster

Doesn't seem to output the `service_ipv6_cidr`.

So you'd have to do the work-around of installing the latest version of the `aws` inside the AMI and add what @marcuz showed:

```
--ip-family ipv6 --service-ipv6-cidr $(aws eks describe-cluster --region=${var.region} --name=${var.cluster_name} --output=text --query 'cluster.{serviceIpv6Cidr: kubernetesNetworkConfig.serviceIpv6Cidr}')
```

This can and should be done by default in the template file that is included with this module if the cluster is in IPv6 mode.

**Comment by bryantbiggs** [comment]
@archoversight unfortunately there isn't anything that this module can do to fix the AMI - we will have to rely on the upstream repo for all AMI settings
